network-api/networkapi/wagtailpages/pagemodels/buyersguide/campaign_page.py

```diff
@@ -0,0 +1,131 @@
+        related_name='donation_modal_relations',
+    )
+    donation_modal = models.ForeignKey(
+        'DonationModal',
+        null=False,
+        blank=False,
+        on_delete=models.CASCADE,
+        help_text='Choose existing or create new donation modal'
+    )
+
+    panels = [
+        SnippetChooserPanel('donation_modal'),
+    ]
+
+    class Meta(TranslatableMixin.Meta, Orderable.Meta):
+        verbose_name = 'Donation Modals'
+        verbose_name_plural = 'Donation Modals'
+
+
+class BuyersGuideCampaignPage(FoundationMetadataPageMixin, Page):
+    parent_page_types = ['BuyersGuideEditorialContentIndexPage']
+    subpage_types: list = []
+    template = 'pages/buyersguide/campaign_page.html'
+
+    cta = models.ForeignKey(
+        'CTA',
+        related_name='buyersguide_campaign_page_for_cta',
+        blank=True,
+        null=True,
+        on_delete=models.SET_NULL,
+        help_text='Choose one of our call-to-action snippets, or create a new one.'
+    )
+    header = models.CharField(
+        max_length=250,
+        blank=True
+    )
+    narrowed_page_content = models.BooleanField(
+        default=False,
+        help_text='For text-heavy pages, turn this on to reduce the overall width of the content on the page.'
+    )
+    body = StreamField(
+        block_types=(('accordion', customblocks.AccordionBlock()),
+                     ('paragraph', RichTextBlock(
+                         features=full_content_rich_text_options,
+                         template='wagtailpages/blocks/rich_text_block.html',
+                     )),
+                     ('card_grid', customblocks.CardGridBlock()),
+                     ('image_grid', customblocks.ImageGridBlock()),
+                     ('iframe', customblocks.iFrameBlock()),
+                     ('image', customblocks.AnnotatedImageBlock()),
+                     ('audio', customblocks.AudioBlock()),
+                     ('image_text', customblocks.ImageTextBlock()),
+                     ('image_text_mini', customblocks.ImageTextMini()),
+                     ('video', customblocks.VideoBlock()),
+                     ('linkbutton', customblocks.LinkButtonBlock()),
+                     ('looping_video', customblocks.LoopingVideoBlock()),
+                     ('pulse_listing', customblocks.PulseProjectList()),
+                     ('single_quote', customblocks.SingleQuoteBlock()),
+                     ('slider', customblocks.FoundationSliderBlock()),
+                     ('spacer', customblocks.BootstrapSpacerBlock()),
+                     ('airtable', customblocks.AirTableBlock()),
+                     ('datawrapper', customblocks.DatawrapperBlock()),
+                     ('typeform', customblocks.TypeformBlock()), )
+    )
+
+    def get_donation_modal_json(self):
+        modals = self.donation_modal_relations.all()
+        modals_json = [m.donation_modal.to_simple_dict() for m in modals]
+        return json.dumps(modals_json)
```

Comment:
What do you think about creating a utility function for this? I guess the same code is needed on the non-buyerguide donation page and the format is required by the JS? I know it's not the third repetition of this pattern but I guess the dependent JS might make it worth it.

Reply:
Hi Tibor, I would vote against this as there is a minor difference between the functions for `CampaignPage`, and `BuyersGuideCampaignPage`.

For `BuyersGuideCampaignPage`, I was able to update the relationship to now use an orderable, similar to what we use for other recent models.

So in this function the name of the relationship has been updated from `donation_modals` to `donation_modal_relations` and line 90 has been updated from 

```
        modals_json = [m.to_simple_dict() for m in modals]
``` 
to

```
        modals_json = [m.donation_modal.to_simple_dict() for m in modals]
```

Reply:
Thanks for expanding on this. I think the difference stems from the fact that the original relation model `DonationModals` also defines that `to_simple_dict` method... which in a weird way defers to the `DonationModeal.to_simple_dict`...